Gui with menubars(WIP)

gautam1168 opened the pull request:
GUI under construction. 

Running the gui
--------------------

1. Navigate to the folder where you have the yaml configuration file and the hdf5 data file.
2. Setup the Qt binding you want to use. (Use only one of the following two lines)

 ```bash
     export QT_API=pyside
     export QT_API=pyqt    
     ``` 

     Then start the ipython console with eventloop integration.

 ```bash 
    ipython --pylab=qt4
     ```

3. import the gui

 ```python
    from tardis import gui
    ```

4. To run the active mode:
 ```python
     win = gui.Tardis(None, 'config.yml', 'data.h5')
     ```
   To run the passive mode you need to have a model already calculated and ready to be displayed:
   
 ```python
    win = gui.Tardis()
    win.show_model(mdl)
    ```

gautam1168 commented:
![guipreview-1](https://cloud.githubusercontent.com/assets/9499740/7491596/578f0a54-f408-11e4-987f-76d8ae518e55.png)

gautam1168 commented:
@wkerzendorf @philkoene, what is Michael's tag?
Hi
I have uploaded a screenshot above. The top-right corner shows a label with model parameters. Does anyone have any suggestions regarding those or the appearance of the plots?

wkerzendorf commented:
It's @mklauser. This Looks very nice! Can we quickly chat with you (I sent
you a message on gitter)
On Wed, May 6, 2015 at 12:38 Gaurav Gautam <notifications@github.com> wrote:

> @wkerzendorf <https://github.com/wkerzendorf> @philkoene
> <https://github.com/philkoene>, what is Michael's tag?
> Hi
> I have uploaded a screenshot above. The top-right corner shows a label
> with model parameters. Does anyone have any suggestions regarding those or
> the appearance of the plots?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/296#issuecomment-99413774>.
>

gautam1168 commented:
I have made some changes to the GUI. Now it can be launched in active or passive mode. The passive mode is the same as the one you have seen. The active mode has an extra tab that will hold the form to edit the config. I have not pushed these yet. I will do that once the form is ready. I am going through the configuration and I see that many fields are mandatory but still have a default value. But why are they mandatory if they have defaults?

gautam1168 commented:
In the configuration template there are fields called type which can take values from a list. Just below that there are what appears to be options that go with each type. Like for convergence_strategy there is damped and  specific. Then there are things like +damped, _damped etc. What do these mean?

gautam1168 commented:
![guipreviewshell](https://cloud.githubusercontent.com/assets/9499740/7531051/4decc7d8-f571-11e4-9a93-a237f176f864.png)
![guipreviewform](https://cloud.githubusercontent.com/assets/9499740/7531055/6fe18860-f571-11e4-9494-c3e8c741623c.png)

gautam1168 commented:
@wkerzendorf @philkoene @mklauser The form looks bad because its incomplete. I have a function that can convert the configuration dictionary into a form. But it still needs some work which I can do once you answer my previous questions. I'll look at the errors and bad smells that the landscape not is complaining about once I am done with the form, at which point the user will be able to modify and recalculate within the GUI. If you want to run this right now then you should recompile the qrc file and import the gui. And then use: gui.Tardis(yamlconfigfile, atomdatahdffile)

wkerzendorf commented:
@gautam1168 @mklauser has suggested using `Qtree` for the pretty hierarchical structure of the configuration. But other than that - very nice work.

wkerzendorf commented:
@gautam1168 I think another nice addition would be a script (maybe called `tardisgui`) that would launch the GUI.

gautam1168 commented:
@wkerzendorf that should be easy enough. I am trying to figure out how to make the tree widget and how to put spinboxes and lineEdits in it. I know that it has to be done by writing custom delegates for the view. I think a tree model with a QColumnView will be ideal. But I'll need some time to go through the documentation of qt. So I think I should have a working form and the script in 2-3 days.

gautam1168 commented:
![guipreview](https://cloud.githubusercontent.com/assets/9499740/7553213/8d9c4d1c-f707-11e4-93c5-e8c8d2c8c22d.png)

gautam1168 commented:
That's the new configuration editor. It allows keyboard navigation using arrow keys. It cannot yet rerun the calculations.

wkerzendorf commented on setup.py:
```diff
@@ -118,3 +118,8 @@
 )
+
+#Compiling the resource file for the gui
+import subprocess
```
Doesn't seem to work on anaconda
```python
Using /Users/wkerzend/anaconda3/envs/tardis-devel/lib/python2.7/site-packages
Finished processing dependencies for tardis-sn==1.0.1
Traceback (most recent call last):
  File "setup.py", line 125, in <module>
    '-o', 'tardis/resource_rc.py'])
  File "/Users/wkerzend/anaconda3/envs/tardis-devel/lib/python2.7/subprocess.py", line 522, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/Users/wkerzend/anaconda3/envs/tardis-devel/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/Users/wkerzend/anaconda3/envs/tardis-devel/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory```

Have a look here how they implement the compilation of `.qrc` files

gautam1168 replied:
I'll check it.

gautam1168 replied:
@wkerzendorf 
I have done what was done in the link you provided. And it seems to be working for me:

----------------------------------------------------------------------------------------------------------------------
 python setup.py develop
Freezing version number to tardis/version.py
running develop
running egg_info
creating tardis_sn.egg-info
writing requirements to tardis_sn.egg-info/requires.txt
writing tardis_sn.egg-info/PKG-INFO
writing top-level names to tardis_sn.egg-info/top_level.txt
writing dependency_links to tardis_sn.egg-info/dependency_links.txt
writing entry points to tardis_sn.egg-info/entry_points.txt
writing manifest file 'tardis_sn.egg-info/SOURCES.txt'
reading manifest file 'tardis_sn.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching 'CHANGES.rst'
warning: no files found matching '*.c' under directory '*.pyx'
warning: no files found matching '*.pxd' under directory '*.pyx'
warning: no files found matching '*.h' under directory '*.pyx'
warning: no files found matching '*' under directory 'licenses'
warning: no files found matching '*' under directory 'cextern'
no previously-included directories found matching 'build'
no previously-included directories found matching 'docs/_build'
no previously-included directories found matching 'docs/api'
warning: no previously-included files matching '*.o' found anywhere in distribution
writing manifest file 'tardis_sn.egg-info/SOURCES.txt'
running build_ext
cythoning tardis/montecarlo/montecarlo.pyx to tardis/montecarlo/montecarlo.c
building 'tardis.montecarlo.montecarlo' extension
creating build
creating build/temp.linux-x86_64-2.7
creating build/temp.linux-x86_64-2.7/tardis
creating build/temp.linux-x86_64-2.7/tardis/montecarlo
creating build/temp.linux-x86_64-2.7/tardis/montecarlo/src
creating build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itardis/montecarlo/src -Itardis/montecarlo/src/randomkit -I/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include -I/home/gaurav/anaconda/envs/mytardisclonenv/include/python2.7 -c tardis/montecarlo/montecarlo.c -o build/temp.linux-x86_64-2.7/tardis/montecarlo/montecarlo.o
In file included from /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1804:0,
                 from /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:17,
                 from /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from tardis/montecarlo/montecarlo.c:250:
/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/__ufunc_api.h:241:1: warning: ‘_import_umath’ defined but not used [-Wunused-function]
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itardis/montecarlo/src -Itardis/montecarlo/src/randomkit -I/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include -I/home/gaurav/anaconda/envs/mytardisclonenv/include/python2.7 -c tardis/montecarlo/src/cmontecarlo.c -o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/cmontecarlo.o
tardis/montecarlo/src/cmontecarlo.c: In function ‘compute_distance2boundary’:
tardis/montecarlo/src/cmontecarlo.c:139:10: warning: unused variable ‘result’ [-Wunused-variable]
tardis/montecarlo/src/cmontecarlo.c: In function ‘compute_distance2line’:
tardis/montecarlo/src/cmontecarlo.c:225:4: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘int64_t’ [-Wformat]
tardis/montecarlo/src/cmontecarlo.c: In function ‘rpacket_init’:
tardis/montecarlo/src/cmontecarlo.c:681:8: warning: unused variable ‘close_line’ [-Wunused-variable]
tardis/montecarlo/src/cmontecarlo.c:672:10: warning: unused variable ‘nu_line’ [-Wunused-variable]
tardis/montecarlo/src/cmontecarlo.c: In function ‘binary_search’:
tardis/montecarlo/src/cmontecarlo.c:109:16: warning: ‘imid’ may be used uninitialized in this function [-Wuninitialized]
tardis/montecarlo/src/cmontecarlo.c: In function ‘montecarlo_compute_distances’:
tardis/montecarlo/src/cmontecarlo.c:915:18: warning: ‘d_line’ may be used uninitialized in this function [-Wuninitialized]
tardis/montecarlo/src/cmontecarlo.c:588:14: note: ‘d_line’ was declared here
tardis/montecarlo/src/cmontecarlo.c: In function ‘get_event_handler’:
tardis/montecarlo/src/cmontecarlo.c:915:18: warning: ‘d_line’ may be used uninitialized in this function [-Wuninitialized]
tardis/montecarlo/src/cmontecarlo.c:588:14: note: ‘d_line’ was declared here
tardis/montecarlo/src/cmontecarlo.c: In function ‘montecarlo_one_packet_loop’:
tardis/montecarlo/src/cmontecarlo.c:915:18: warning: ‘d_line’ may be used uninitialized in this function [-Wuninitialized]
tardis/montecarlo/src/cmontecarlo.c:588:14: note: ‘d_line’ was declared here
tardis/montecarlo/src/cmontecarlo.c: In function ‘montecarlo_one_packet’:
tardis/montecarlo/src/cmontecarlo.c:392:3: warning: ‘reabsorbed’ may be used uninitialized in this function [-Wuninitialized]
tardis/montecarlo/src/cmontecarlo.c:388:22: warning: ‘weight’ may be used uninitialized in this function [-Wuninitialized]
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itardis/montecarlo/src -Itardis/montecarlo/src/randomkit -I/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include -I/home/gaurav/anaconda/envs/mytardisclonenv/include/python2.7 -c tardis/montecarlo/src/randomkit/rk_isaac.c -o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_isaac.o
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itardis/montecarlo/src -Itardis/montecarlo/src/randomkit -I/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include -I/home/gaurav/anaconda/envs/mytardisclonenv/include/python2.7 -c tardis/montecarlo/src/randomkit/rk_mt.c -o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_mt.o
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itardis/montecarlo/src -Itardis/montecarlo/src/randomkit -I/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include -I/home/gaurav/anaconda/envs/mytardisclonenv/include/python2.7 -c tardis/montecarlo/src/randomkit/rk_sobol.c -o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_sobol.o
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Itardis/montecarlo/src -Itardis/montecarlo/src/randomkit -I/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include -I/home/gaurav/anaconda/envs/mytardisclonenv/include/python2.7 -c tardis/montecarlo/src/randomkit/rk_primitive.c -o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_primitive.o
gcc -pthread -shared build/temp.linux-x86_64-2.7/tardis/montecarlo/montecarlo.o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/cmontecarlo.o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_isaac.o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_mt.o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_sobol.o build/temp.linux-x86_64-2.7/tardis/montecarlo/src/randomkit/rk_primitive.o -L/home/gaurav/anaconda/envs/mytardisclonenv/lib -lpython2.7 -o /home/gaurav/GSoC/installationtesting/tardis/tardis/montecarlo/montecarlo.so
cythoning tardis/macro_atom.pyx to tardis/macro_atom.c
building 'tardis.macro_atom' extension
gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include -I/home/gaurav/anaconda/envs/mytardisclonenv/include/python2.7 -c tardis/macro_atom.c -o build/temp.linux-x86_64-2.7/tardis/macro_atom.o
In file included from /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1804:0,
                 from /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:17,
                 from /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from tardis/macro_atom.c:250:
/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/__multiarray_api.h:1629:1: warning: ‘_import_array’ defined but not used [-Wunused-function]
/home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/numpy/core/include/numpy/__ufunc_api.h:241:1: warning: ‘_import_umath’ defined but not used [-Wunused-function]
gcc -pthread -shared build/temp.linux-x86_64-2.7/tardis/macro_atom.o -L/home/gaurav/anaconda/envs/mytardisclonenv/lib -lpython2.7 -o /home/gaurav/GSoC/installationtesting/tardis/tardis/macro_atom.so
Creating /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages/tardis-sn.egg-link (link to .)
Removing tardis-sn 1.0.1 from easy-install.pth file
Adding tardis-sn 1.0.1 to easy-install.pth file
Installing tardis script to /home/gaurav/anaconda/envs/mytardisclonenv/bin

Installed /home/gaurav/GSoC/installationtesting/tardis
Processing dependencies for tardis-sn==1.0.1
Searching for astropy==1.0
Best match: astropy 1.0
Adding astropy 1.0 to easy-install.pth file

Using /home/gaurav/anaconda/envs/mytardisclonenv/lib/python2.7/site-packages
Finished processing dependencies for tardis-sn==1.0.1
-----------------------------------------------------------------------------------------------------------------------

Your error seems to be saying that a file or directory is missing. Did you clone from 
git@github.com:gautam1168/tardis.git ?

wkerzendorf replied:
I think we should use the structure that `glue-viz` uses - that is a proper use of the setup structure and ensures that things are working.

gautam1168 commented:
@mklauser Can you tell me how to display icons for qactions without using a qrc resource file?

gautam1168 commented:
@wkerzendorf 

I asked Chris Beaumont and from his code I see that he is using the full path to the icon file. But the method that he is using is not working for me. The path that his function generates starts from the installation directory of his package but what I need is a path starting at root. I have found a way of doing this which is shorter than his way but looks like a hack. (On the plus side, I can't think of a reason why this will not work on a mac unless mac uses a different separator than '/' in paths).  Here is what I do inside the gui:

import tardis
path = tardis.__file__    
path = path.split('/')
path.pop()
path.append('images')
path = '/'.join(path)
path = path+'/'

So path is /home/gaurav/GSoC/mytardisclone/tardis/tardis/images/. Please see if you get this result on a mac.

Now I can use this to display icons without the qrc files. I am pushing this now.

wkerzendorf commented:
@gautam1168 a couple of things. 

If you want to show code in a github comment use trip backticks to make sure it is formatted well. That makes it much easier to read. 

For dealing with paths you should use `os.path`. Here is an example to get a path to tardis'es data directory:

```python
import os
import tardis

data_path = os.path.join(tardis.__path__[0], 'data')
```
good luck.

wkerzendorf commented:
@gautam1168 can you give a quick step-by-step guide in the beginning of this PR how to run it.

gautam1168 commented:
Should I write it on the thread at github?
On May 17, 2015 4:33 PM, "Wolfgang Kerzendorf" <notifications@github.com>
wrote:
>
> @gautam1168 can you give a quick step-by-step guide in the beginning of
this PR how to run it.
>
> —
> Reply to this email directly or view it on GitHub.

wkerzendorf commented:
Just edit the description of this PR - at the very top

gautam1168 commented:
Tell me if it is working.

wkerzendorf commented:
very nice!

wkerzendorf commented:
I wonder if we want to split this up into multiple windows. what do you think?

wkerzendorf commented:
on the right hand side where we have rad.temp and ws - we should make more columns - density and abundances so O, si, ...

gautam1168 commented:
The window is inherited from QMainWindow and its central widget it a tabbed widget. That means that if we want to we can put more tabs just like the config editor. Is that what you mean by multiple windows? Making more columns should be easy since its Qt that is making them. One just has to provide the data to the QTableModel.

wkerzendorf commented:
no I'm thinking of having multiple independent windows

gautam1168 commented:
But more importantly, the config editor is not ready yet. I need to make it able to rerun the calculations. Right now the recalculate button does nothing. For that I need to go through the config_validator and figure out how it works and how to use that in this case.

gautam1168 commented:
So do you just want to display two different windows together or do you want these windows to be launchable separately? Because if you want the former then I think looking into dockable widgets will be fruitful. But the second option is more like having two unrelated gui's

wkerzendorf commented:
I understand - I think it would be great to have all these things as separate windows. I think the first option would be good.

gautam1168 commented:
So the config editor in one window and the display widgets in another. Like matlab?

wkerzendorf commented:
I think so - let me be perfectly honest with you: You are doing frontier work - we do not quite know what we want. so there might be some trial/error :wink:

gautam1168 commented:
That is fine since I don't have deadlines :)

wkerzendorf commented:
@gautam1168 I think the next step is to clean this code, disable the Parameter input for now. And add the bit at the top to the documentation. Thanks!

gautam1168 commented:
Hi Wolfgang,

I'll get to it as soon as I can. Hopefully I can start devoting some time
to tardis again tomorrow. I am working out some theory for my other project
and its sucking up all my time.
On May 24, 2015 4:50 PM, "Wolfgang Kerzendorf" <notifications@github.com>
wrote:

> @gautam1168 <https://github.com/gautam1168> I think the next step is to
> clean this code, disable the Parameter input for now. And add the bit at
> the top to the documentation. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/296#issuecomment-105003094>.
>

wkerzendorf commented:
@gautam1168 - there seems to be a problem with the line-id plot (also discovered @unoebauer) . Do you know where this is? It would be very helpful to fix that before merging this.  Thanks!

gautam1168 commented:
What is the problem? Is there an issue created by @unoebauer? I can't find it.

wkerzendorf commented:
@gautam1168 - we just checked it. it seems to work for the new instructions you gave. WE'll try it out. Sorry we didn't create a new issue. 

@kaushik94 - can you quickly go over and have a look at the coding style? (For your info @gautam1168 - @kaushik94 is helping us to make sure that we have a better coding style and will help to improve things by comparing to the PEP8 standard; I'll let him explain what this is.)

gautam1168 commented:
Yes I was looking at that myself but it will be very helpful if someone
explained it to me. @kaushik94 tell me what's a good time for you and we
can do a hangout. I am not sure a video one will be easy with my internet
but we could chat or talk on the phone. Looking forward to talking to you.
On May 26, 2015 8:51 PM, "Wolfgang Kerzendorf" <notifications@github.com>
wrote:

> @gautam1168 <https://github.com/gautam1168> - we just checked it. it
> seems to work for the new instructions you gave. WE'll try it out. Sorry we
> didn't create a new issue.
>
> @kaushik94 <https://github.com/kaushik94> - can you quickly go over and
> have a look at the coding style? (For your info @gautam1168
> <https://github.com/gautam1168> - @kaushik94
> <https://github.com/kaushik94> is helping us to make sure that we have a
> better coding style and will help to improve things by comparing to the
> PEP8 standard; I'll let him explain what this is.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/296#issuecomment-105564700>.
>

unoebauer commented on tardis/gui.py:
```diff
@@ -1,5 +1,12 @@
 import numpy as np
 import matplotlib
+matplotlib.style.use('fivethirtyeight')
```
The style packet has been added to matplotlib starting with version 1.4:

http://matplotlib.org/users/whats_new.html#style-package-added

To make the gui work also with older versions, maybe add a check which only uses this style when matplotlib >= 1.4 is available.

gautam1168 replied:
Sure I can add a check.
On May 27, 2015 3:47 PM, "unoebauer" <notifications@github.com> wrote:

> In tardis/gui.py
> <https://github.com/tardis-sn/tardis/pull/296#discussion_r31119776>:
>
> > @@ -1,5 +1,12 @@
> >  import numpy as np
> >  import matplotlib
> > +matplotlib.style.use('fivethirtyeight')
>
> The style packet has been added to matplotlib starting with version 1.4:
>
> http://matplotlib.org/users/whats_new.html#style-package-added
>
> To make the gui work also with older versions, maybe add a check which
> only uses this style when matplotlib >= 1.4 is available.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/296/files#r31119776>.
>

kaushik94 commented:
Hi @gautam1168 , firstly good effort there. There are too many files to review here so will take some time  may be if its not urgent ?. And about the hangout, first create a doc and let me know what we need to discuss and I'll may be brush up or prepare. That won't waste your time. What say ?

gautam1168 commented:
Okay I'll send you a doc.
On May 28, 2015 12:45 AM, "Kaushik Varanasi" <notifications@github.com>
wrote:

> Hi @gautam1168 <https://github.com/gautam1168> , firstly good effort
> there. There are too many files to review here so will take some time may
> be if its not urgent ?. And about the hangout, first create a doc and let
> me know what we need to discuss and I'll may be brush up or prepare. That
> won't waste your time. What say ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/296#issuecomment-106037301>.
>

gautam1168 commented:
@kaushik94 
Hi !,
Can you tell me if I am missing something? So far the following is what I have done:

1. Added docstring to the gui.py file. (This is the only file I ever edit.)
2. Added docstrings to all classes
3. Adding docstrings to all methods (in progress)
4. Arranged imports as specified in PEP
5. Cropped lines to be below 80 characters (with very few exceptions).

gui.py has class definitions and nothing else. It is imported and its classes are instantiated elsewhere. Is this information enough for you to tell me if I am missing something?

kaushik94 commented on tardis/gui.py:
```diff
@@ -1,6 +1,39 @@
+Functions:
+---------
+    show_model(model) - A public method of Tardis(not gui itself) to take a 
+    model object and show its data.
```
An indentation here. Take a look at http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_google.html

gautam1168 replied:
But the example you have linked has no indentation for module level docstring. Where should the indent be?

kaushik94 replied:
```
Attributes:
  module_level_variable (int): Module level variables may be documented in
    either the ``Attributes`` section of the module docstring, or in an
    inline docstring immediately following the variable.
```
see the indent after the `modeule_level_variable` line ?

gautam1168 replied:
Oh I see! Thanks for being patient. I am converting these to numpy style
now. And I'll use sphinx to test if the doc builds. But I might still need
to bother you a few (and by a few I mean a lot :) more times.
On Jun 1, 2015 8:18 PM, "Kaushik Varanasi" <notifications@github.com> wrote:

> In tardis/gui.py
> <https://github.com/tardis-sn/tardis/pull/296#discussion_r31432286>:
>
> > +            export QT_API=pyqt
> > +    2. Start the IPython console with eventloop integration ::
> > +            ipython --pylab=qt4
> > +    3. Display your model::
> > +            from tardis import gui
> > +            win = gui.Tardis()
> > +            win.show_model(mdl)
> > +
> > +Classes:
> > +-------
> > +    Tardis - The top level QMainWindow widget.
> > +
> > +Functions:
> > +---------
> > +    show_model(model) - A public method of Tardis(not gui itself) to take a
> > +    model object and show its data.
>
> Attributes:
>   module_level_variable (int): Module level variables may be documented in
>     either the ``Attributes`` section of the module docstring, or in an
>     inline docstring immediately following the variable.
>
> see the indent after the modeule_level_variable line ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/296/files#r31432286>.
>

kaushik94 commented on tardis/gui.py:
```diff
@@ -1,6 +1,39 @@
+"""
+import os
+from pkg_resources import parse_version
+import exceptions
```
Try to keep all imports at the top and imports convention should be something like #308. A small discussion regarding coding conventions we had.

gautam1168 replied:
But the numpy example shows imports after the module level docstring.

https://github.com/numpy/numpy/blob/master/doc/example.py

kaushik94 commented on tardis/gui.py:
```diff
@@ -8,32 +41,61 @@
-    
+else:
+    raise ImportError('QT_API was not set! Please exit the IPython console'+
+        ' and at the bash prompt use : \n\n export QT_API=pyside \n or \n'+
```
This is bad. Don't use `+` just to extend the lines. Try doing something like this
```
This is how you \
write mulit-line \
functions

```

gautam1168 replied:
Are you sure about that? I remember reading that \ was not the preferred way of line breaking and one should use the implied line breaks inside (), [] and {} whenever possible and use \ only as a last resort. I am not an expert on PEP though. Just wanted to confirm before I changed these.

kaushik94 replied:
Yes `\` is the last resort but `+` is bad ! If you have a better solution to make it multiline then its good but `+` should be avoided for the multiline purpose as far as I know.

gautam1168 replied:
It turns out that adjecent strings are automatically concatenated. So my +'s were redundant and the \'s are not required. Serendipity :D

kaushik94 commented on tardis/gui.py:
```diff
@@ -8,32 +41,61 @@
+else:
+    print "Please upgrade matplotlib to a version >=1.4 for best results!"
+matplotlib.rcParams['font.family']='serif'
+matplotlib.rcParams['font.size']=10.0
```
I Think there should be a space around `=` on both sides.

kaushik94 commented on tardis/gui.py:
```diff
@@ -8,32 +41,61 @@
+        2.show_model 
+          Take the tardis model object and give it to the Tardis main window 
+          object for display, and call the show method to paint the main 
+          window.
```
A new line here. Similarly to all other doc strings. Read this carefully https://www.python.org/dev/peps/pep-0257/

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+        #path to icons folder
+        self.path = os.path.join(tardis.__path__[0],'images')  
+
+        #Check if configuration file was provided
```
A space after `#`

gautam1168 replied:
Removed

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+        
+        #In case of active mode
+        if self.mode == 'active':
+            #Disabled currently
```
Please delete these lines if not used.

gautam1168 replied:
I will be uncommenting them in the next PR. Is it fine to leave them then?

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+        self.showMaximized()
+
+    #Note to self: Add slot decorator decorator. And try to get all the
+    #view changer slots into a single function
```
Don't add personal notes. You might forget to delete them before merging. Try to use the checkboxes like I did in #272

gautam1168 replied:
done

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+            
+            #Then append siblings of current selection for all type nodes to
+            #disabled nodes
+            
```
Delete this new line. Comments should stick to what they are describing.

gautam1168 replied:
done

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+        return QtCore.Qt.ItemIsEnabled | QtCore.Qt.ItemIsSelectable        
+
+    def getItem(self, index):
+        if index.isValid():
```
I think it can be done more pythonic
```
return index.internalPointer() if index.isValid() and index.internalPointer() else self.root
```

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+        parentItem = self.getItem(parent)
+        self.beginInsertRows(parent, position, position + rows - 1)
+        success = parentItem.insertChildren(position, rows,
+                self.rootItem.columnCount())
```
I'm not sure but I think there should be another tab here. Just check.

gautam1168 replied:
done

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+
+class TreeDelegate(QtGui.QStyledItemDelegate):
+    """Create a custom delegate to modify the columnview that displays the 
+    TreeModel."""
```
The """ should go in the next line. So it should be
```
"""Create a custom delegate to modify the columnview that displays the 
TreeModel.
*new line here*
"""
```

gautam1168 replied:
done

kaushik94 commented on tardis/gui.py:
```diff
@@ -43,68 +105,739 @@ def __init__(self, parent=None):
+                     .format(self.model.iterations_max_requested, 
+                        self.model.iterations_executed,
+                        '<font color="green"><b>True</b></font>' if 
+                        self.model.converged else 
```
These 3 lines are unnecessary. Just do
```
'<font color="green"><b>'+ str(self.model.converged) +'</b></font>'
```

wkerzendorf replied:
very close. you should use string formatting. so do: `"<font color='green'><b>{0}</b></font>".format(self.model.converged)

gautam1168 replied:
Well both of your solutions won't work because I added this statement not to change the True to False but the color of the string 'True' or 'False'. False is red and True is green. But in both of your suggested solutions the string will always be green.

@wkerzendorf, if I really want to do a formatting of this string I have to do something like:

```python
    truthstring = "<font color={}>{}</font>"
    if (model.converged):
        truthstring.format('green','True')
    else:
        truthstring.format('red','False')
"blah blah".format("blah1","blah2", truthstring)
```
I think that is needlessly complicated. Will it be more pleasing to the eyes if I do something like:

```python
    truth1 = "<font color=green>True</font>"
    truth2 = "<font color=red>False</font>"
    "blah blah".format("blah1","blah2", truth1 if model.converged else truth2)
```
I still feel like the code already written is the clearest but I don't know what the PEP says about these things. So which one should I put in.

kaushik94 replied:
@gautam1168 I din't see the red-green thing. Let it be the way it is.

kaushik94 commented:
I din't check it throughly but a few comment regarding coding and commenting. Please read 
https://www.python.org/dev/peps/pep-0257/
http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_google.html

And also try using some software like http://sphinx-doc.org/

wkerzendorf commented:
@kaushik94 @gautam1168 I should have mentioned - we use numpydoc docstrings: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt.

Thanks to you both for looking into this!

kaushik94 commented on tardis/gui.py:
```diff
@@ -8,32 +41,61 @@
+    
+    Public methods:
+    --------------
+        1. __init__
```
I don't think you should use numbering in docstrings though I am not so sure.

gautam1168 replied:
I'll find out if that is allowed.

gautam1168 commented:
@wkerzendorf Okay I am going to follow the guide you linked. I was wondering if building with sphinx is a sufficient condition to check if the style is right.

gautam1168 commented:
@kaushik94 @wkerzendorf Its done.

wkerzendorf commented:
It looks nice! Thank you.

wkerzendorf commented:
@gautam1168 next PR we need to make sure we put this in the documentation

gautam1168 commented:
So what comes next?
On Jun 4, 2015 8:36 PM, "Wolfgang Kerzendorf" <notifications@github.com>
wrote:

> It looks nice! Thank you.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/296#issuecomment-108927960>.
>

wkerzendorf commented:
the next quick thing is to put this in the documentation. go to the docs folder and make a gui.rst file. 
Then copy what you have in the beginning of this file there. Make a PR and I'll help you further.
